### ISSUE TITLE
codegen export to C | superfluous "1.0" as in x*1.0/y for x/y

### ISSUE THREAD

**Issue by Audrius-St** [sympy/sympy#21815]
Hello, 

I have been using sympy codegen to export sympy code to C. 
In a new project, I've encountered the following issue of the form

`rhs_expr = x*(r*r*r)**(-1) `

is exported by codegen as 

`y6 = x*1.0/(r*r*r);`

The "1.0" in the above expression has two issues for me:

1. It is superfluous

2.  In C I am using an extended precision library that requires a specialized data type.

Re the SymPy mailing list discussion:

https://groups.google.com/g/sympy/c/wzz4DVWMRF8

"The 1.0 is currently hard-coded in the C code printer
https://github.com/sympy/sympy/blob/aa538a209d66b45eb73d9579fa0d088fe0adf220/sympy/printing/c.py#L284."

**Comment by asmeurer** [comment]
As I said on the mailing list:

> However, I think this should be fixed in SymPy itself. If you look at
the other printers, like the normal str printer, it doesn't print x/y
as x*1/y. If you look at the code (see sympy/printing/str.py), it's
handled more intelligently in the _print_Mul printer. I think the code
printer(s) should also handle Mul more intelligently, so that they
don't print superfluous 1.0s. It also makes sense to me that when it
does print 1.0, that it instead recursively calls _print_Float. That
way if you want to handle floats in a special way in the printer, you
can just redefine _print_Float.

**Comment by bjodah** [comment]
I agree that the current output is suboptimal. My guess is that it was made like this to force floating point division when variables are declared as e.g. `int`.

@Audrius-St if you can/want to share I'm interested to hear what extended precision library you're working with in C which doesn't implicitly convert 1.0 losslessly into the underlying type.

**Comment by Audrius-St** [comment]
> @Audrius-St if you can/want to share I'm interested to hear what extended precision library you're working with in C which doesn't implicitly convert 1.0 losslessly into the underlying type.

Paper: "CAMPARY: Cuda Multiple Precision Arithmetic Library and Applications"

https://link.springer.com/chapter/10.1007/978-3-319-42432-3_29

Thesis: "Towards fast and certified multiple-precision librairies"

https://hal.archives-ouvertes.fr/tel-01534090/

Code: https://homepages.laas.fr/mmjoldes/campary/

I'm currently using the CPU version which performs calculations in my application in double double extended precision between  10 to 20 times faster than the equivalent in MPFR, specifically,  the user-friendly MPFR C++ wrapper by Pavel Holoborodko et al.
while attaining Brouwer's Law.

In my applications, I have high degree multivariate polynomials with rational coefficients, the numerators and denominators of which I declare static const <campary datatype> based on "Optimizing software in C++" by Agner Fog.

**Comment by bjodah** [comment]
Thanks! OK, so rather C++ than C then? So if you have explicit constructors, I guess you can also overload `_print_Float` in your custom printer subclass, to have it emit wrapped number literals: `T{...}` (where `T` is your class).

But as @asmeurer pointed out, we should change the printing in the base class here (the C-code printer) not to emit `1.0`.

I've worked some with the qd library by David H. Bailey et al. and boost::multiprecision, but cool see things progressing and finding it's way to the GPU!

**Comment by asmeurer** [comment]
> I agree that the current output is suboptimal. My guess is that it was made like this to force floating point division when variables are declared as e.g. int.

I hadn't considered that. Don't the printers generally assume everything is going to be float? It does seem possible though that changing this could break someone's code.
